Add .NET Standard 2.0 as target

pgrawehr opened the pull request:
I was attempting to use a binding in a project that still uses .NET Framework (and cannot be migrated just yet) and found that this was not possible. This adds cross-build support for most bindings with only minor tweaks. 


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/iot/pull/1530)

joperezr commented on src/devices/Directory.Build.props:
```diff
@@ -3,6 +3,7 @@
   <Import Project="../../eng/Versions.props" />
   <Import Project="../../eng/Analyzers.props" />
   <PropertyGroup>
+    <TargetFrameworksBase>net5.0;netcoreapp2.1;net48;</TargetFrameworksBase>
```
I don't think that adding a net48 configuration is the right thing to do here, instead we should try to add a netstandard2.0 configuration to all projects that support it, and if the build works fine, we can use that instead of the netcoreapp2.1 configuration in case there are no differences between netcoreapp2.1 and netsandard2.0. We already started doing this a while back, but the PR was never completed so perhaps this is a good oportunity to take care of it.

Main advantage of using netstandard2.0 as opposed to net48, is that with it it will be actually making the bindings runnable accross other platforms like Mono, Xamarin, and UWP which will probably be more handy than a net48 configuration.

pgrawehr replied:
I have tried that, but it didn't work. But based on what I have done now, it seems it could actually work. Maybe I'll need your help in getting the packaging stuff right - I can get it to compile, but the way the main Iot.Device.Bindings.dll is built gives me a bit of a headache.

joperezr replied:
Some time ago I merged a PR in Iot.Device.Bindings that will basically get the best target framework for all of the devices for each TFM that Iot.Device.Bindings compiles for, and will try to build  those particular sources for it so in theory this shouldn't be to hard to do now. Work will mostly involve adding netstandard2.0 configs on all devices that support it, and add a netstandard2.0 config to iot.device.bindings. Finally we will just have to add some exclusions into iot.device.bidnings for those devices that won't support netstandard2.0.

joperezr commented on src/devices/Directory.Build.props:
```diff
@@ -3,6 +3,10 @@
+    <!-- Most projects should use this value to build for all possible target frameworks -->
+    <TargetFrameworksBase>net5.0;netcoreapp2.1;netstandard2.0;</TargetFrameworksBase>
+    <!-- If .NET Standard is not supported (and adding support is not useful) use this instead -->
+    <TargetFrameworksReduced>net5.0;netcoreapp2.1</TargetFrameworksReduced>
```
NIT: I would define this one first and then have the one that has netstandard2.0 too be based on this one, that way when we move to target something else the update only goes in one line. Also, I think I would remove these to be more like: DeviceBindingsDefaultTFMs and DeviceBindingsNetCoreTFMs

joperezr commented on src/devices/Common/Iot/Device/Graphics/LcdCharacterEncoding.cs:
```diff
@@ -102,7 +102,7 @@ public override int GetBytes(char[] chars, int charIndex, int charCount, byte[]
                 else
                 {
-                    bytes[byteIndex] = _characterMapping.GetValueOrDefault(_unknownLetter);
+                    bytes[byteIndex] = _characterMapping[_unknownLetter];
```
Should we add a check in case this index doesn't exist? that is what GetValueOrDefault used to do

pgrawehr replied:
Yes. There is already a check in the factory that typically generates this class, but since the ctor is public, I added a validation there. The "unknown" letter must exist.

joperezr commented on src/devices/Common/MathExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <summary>
+        /// Returns val, limited to the range min-max (inclusive)
+        /// </summary>
+        public static double Clamp(double val, double min, double max)
```
It doesn't need to happen now but it would be good to add a few unit tests to these members to make sure they behave as expected.

pgrawehr replied:
Done

joperezr commented on src/devices/HardwareMonitor/OpenHardwareMonitor.cs:
```diff
@@ -657,6 +670,11 @@ public Sensor(ManagementObject instance, string name, string identifier, string?
+            /// <summary>
+            /// The instance ID. Used to retrieve a new value for the same sensor (repeatedly obtaining the same value from a sensor instance returns the same value)
+            /// </summary>
+            internal int InstanceId { get; }
```
Looks like some of the changes on this binding are not related to just the addition of the configurations. Can we for now only include the changes related to netstandard2.0 configs and then send out a follow up for the HardwareMonitor changes? It just m akes it easier to audit these type of changes in the future.

joperezr commented on src/devices/MotorHat/DCMotor3Pwm.cs:
```diff
@@ -33,7 +33,7 @@ public override double Speed
             {
                 // Make sure the speed is between -1 and 1
-                _speed = Math.Clamp(value, -1, 1);
+                _speed = value < -1 ? -1 : (value > 1) ? 1 : value;
```
Why not use the Extension you added?

pgrawehr replied:
I sometimes just recoded it, to avoid having to add the extra project reference everyhwere. Either way is fine for me.

joperezr commented on src/devices/Pca9685/Pca9685.cs:
```diff
@@ -296,7 +296,17 @@ private void SetOnOffTimeAllChannels(ushort on, ushort off)
         private byte FrequencyToPrescale(double freqHz)
         {
             var desiredPrescale = Math.Round(ClockFrequency / 4096 / freqHz - 1);
-            return (byte)Math.Clamp(desiredPrescale, byte.MinValue, byte.MaxValue);
```
Same here, should we use your extension instead?

joperezr commented:
Left a few comments and took the liberty to fix the non-Windows build issues, but this is mostly ready to go once the last few comments are addressed. Thanks for taking care of this @pgrawehr

pgrawehr commented:
Removed the commits that don't belong here and addressed the other findings. 

I have a small problem: I would like to update the DeviceBindingTemplate, too, but that one has not the same Directory.build.props, as it is completely outside of the device tree. Since there are a few similar changes I would like to make to the project files and the template (I want to clean up the test projects as well, since the used versions of xunit and Moq are inconsistent) it would be good if we could just completely exclude that project from build.

joperezr commented on src/devices/Directory.Build.props:
```diff
@@ -12,6 +14,7 @@
     <LangVersion>9</LangVersion>
     <Nullable>enable</Nullable>
   </PropertyGroup>
+  
```
```suggestion
```
NIT: remove extra line

joperezr commented on src/devices/BoardLed/BoardLed.cs:
```diff
@@ -119,7 +119,16 @@ private int GetMaxBrightness()
         private void SetBrightness(int value)
         {
-            value = Math.Clamp(value, 0, 255);
+            if (value < 0)
```
NIT: Use your extension method instead.

joperezr commented on src/devices/Common/FrameworkCompatibilityExtensions.cs:
```diff
@@ -0,0 +1,72 @@
+using System.Threading.Tasks;
+
+#pragma warning disable CS1591 // Missing documentation (these methods are mostly hidden from the user)
+#if NETSTANDARD2_0
```
I disagree here. Having IfDefs just create maintainability burden (we've had this happen MANY times in dotnet/runtime repo) even when the defines are comming from the SDK, so whenever they are easily avoidable that is preferred. VS has a feature for selecting the configuration you are working with and that will automatically show/hide files that should be excluded. That is the pattern that we have followed with the rest of the repo, so I would rather just exclude the file at least for consistency-sake. If we find that the VS experience is not there for this scenario, we can then either file a request for VS to add support for this, or (in the worst case scenario) move the  whole repo to use this pattern, but I think that what should be the driver here is consistency.

joperezr commented on src/devices/SocketCan/CanRaw.cs:
```diff
@@ -55,11 +56,10 @@ public void WriteFrame(ReadOnlySpan<byte> data, CanId id)
-            ReadOnlySpan<CanFrame> frameSpan = MemoryMarshal.CreateReadOnlySpan(ref frame, 1);
-            ReadOnlySpan<byte> buff = MemoryMarshal.AsBytes(frameSpan);
-            Interop.Write(_handle, buff);
+                byte* buff = (byte*)&frame;
```
Where you able to test this? Is the problem here that Netstandard2.0 doesn't have these MemoryMarshal? Aren't these methods available through the System.Memory package?

pgrawehr replied:
No, I could not test this, but I'm quite certain that it's equivalent. I did not look for an implementation of these methods because the changed code is simpler than the old. This converted to two different ReadOnlySpan types and finally to a byte*, while you can just directly take the address of the frame.

joperezr replied:
Ok, I checked and they don't seem to be available on System.Memory package for NS2.0. To be honest, it feels a bit weird to change this when we haven't tested it at all even when we have high confidence on the change, but for now let's keep it and if folks using the binding find an issue here they can log the issue and we can investigate then.

pgrawehr replied:
The binding was originally provided by @krwq I think. So maybe he has the hardware available to test it?

joperezr commented on src/devices/UFireIse/UFireIse.cs:
```diff
@@ -310,6 +310,12 @@ private float ReadFloat(Register register)
+            byte[] bytes = BitConverter.GetBytes(data);
+            ChangeRegister(register);
+            _device.Write(bytes);
+            DelayHelper.DelayMilliseconds(IseCommunicationDelay, allowThreadYield: true);
```
This is no longer throwing the exceptions right?

pgrawehr replied:
True, but that exception cannot really happen. The only situation where `TryWriteBytes` returns false is if there's not enough room in the buffer.

joperezr commented on src/devices/Common/MathExtensions.cs:
```diff
@@ -0,0 +1,130 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static double Clamp(double val, double min, double max)
+        {
+#if NETSTANDARD2_0
```
While I'm fine with this for now, an alternative would be to have two implementations of this file that are conditionally compiled, one for the downlevel implementation, and the other one that just forwards the calls to Math to avoid all these ifdefs.

joperezr reviewed (approved):
Left some last few comments but this is ready to go once those are answered/addressed.

joperezr commented on src/devices/Common/FrameworkCompatibilityExtensions.cs:
```diff
@@ -0,0 +1,86 @@
+using System.Threading.Tasks;
+
+#pragma warning disable CS1591 // Missing documentation (these methods are mostly hidden from the user)
+#if NETSTANDARD2_0
```
Looks like you still haven't removed this Ifdef and instead conditionally compile the file. Same goes for IsExternalInit. I think that is the last remaining feedback I have and this should be good to go.
